pkg/controller/syncsetinstance/syncsetinstance_controller.go

```diff
@@ -548,6 +557,65 @@ func (r *ReconcileSyncSetInstance) applySyncSetResources(ssi *hivev1.SyncSetInst
 	return nil
 }
 
+func (r *ReconcileSyncSetInstance) reconcileDeletedSyncSetResources(applyMode hivev1.SyncSetResourceApplyMode, dynamicClient dynamic.Interface, existingStatusList, newStatusList []hivev1.SyncStatus, err error, ssiLog log.FieldLogger) ([]hivev1.SyncStatus, error) {
+	ssiLog.Debugf("reconciling syncset resources, existing: %d, actual: %d", len(existingStatusList), len(newStatusList))
+	if applyMode == "" || applyMode == hivev1.UpsertResourceApplyMode {
+		ssiLog.Debugf("apply mode is upsert, syncset status will be updated")
+		return newStatusList, nil
+	}
+	deletedStatusList := []hivev1.SyncStatus{}
+	deletedStatusIndices := []int{}
+	for i, existingStatus := range existingStatusList {
+		found := false
+		for _, newStatus := range newStatusList {
+			if existingStatus.Name == newStatus.Name &&
+				existingStatus.Namespace == newStatus.Namespace &&
+				existingStatus.APIVersion == newStatus.APIVersion &&
+				existingStatus.Kind == newStatus.Kind {
+				found = true
+				break
+			}
+		}
+		if !found {
+			ssiLog.WithField("resource", fmt.Sprintf("%s/%s", existingStatus.Namespace, existingStatus.Name)).
+				WithField("apiversion", existingStatus.APIVersion).
+				WithField("kind", existingStatus.Kind).Debug("resource not found in updated status, will queue up for deletion")
+			deletedStatusList = append(deletedStatusList, existingStatus)
+			deletedStatusIndices = append(deletedStatusIndices, i)
+		}
+	}
+
+	// If an error occurred applying resources, do not delete yet
+	if err != nil {
+		ssiLog.Debugf("an error occurred applying resources, will preserve all syncset status items")
+		return append(newStatusList, deletedStatusList...), nil
+	}
+
+	for i, deletedStatus := range deletedStatusList {
+		itemLog := ssiLog.WithField("resource", fmt.Sprintf("%s/%s", deletedStatus.Namespace, deletedStatus.Name)).
+			WithField("apiversion", deletedStatus.APIVersion).
+			WithField("kind", deletedStatus.Kind)
+		gv, err := schema.ParseGroupVersion(deletedStatus.APIVersion)
+		if err != nil {
+			return nil, err
+		}
+		gvr := gv.WithResource(deletedStatus.Resource)
+		itemLog.Debug("deleting resource")
+		err = dynamicClient.Resource(gvr).Namespace(deletedStatus.Namespace).Delete(deletedStatus.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				itemLog.WithError(err).Error("error deleting resource")
+				index := deletedStatusIndices[i]
+				existingStatusList[index].Conditions = r.setDeletionFailedSyncCondition(existingStatusList[index].Conditions, err)
```

Comment:
If there is an error that is not NotFound, then the status for that secret needs to be appended to `newStatusList` rather than updating the condition in the status in `existingStatusList`.

Reply:
I've updated this to add the `deletedStatus` back into the `newStatusList` but I'd like to also add a test for this.